### PR TITLE
Qe 356 rest api json error handlers

### PR DESCRIPTION
**Pull request by kevin-foster-uk**
Yii is configured to display a standard error page in response to unhandled exceptions and fatal errors. When making REST API requests the client expects an JSON response so the standard HTML error page need to be replaced with a JSON response.

### CODE REVIEW

File: application/core/LSYii_Application.php [hunk 1]
```diff
@@ -717,4 +717,93 @@ public function getPublicBaseUrl($absolute = false)
...
+
+        return $acceptsJson || $isRestUrl;
+    }
+
+	/**
+	 * Handles uncaught PHP exceptions.
+	 *
+	 * @param Exception $exception exception that is not caught
+	 */
+    public function handleException($exception)
```
Comment (olleharstedt) [inline application/core/LSYii_Application.php:744]: Hm isn't this configured in config.php? https://www.yiiframework.com/doc/guide/1.1/en/topics.error
Or you can't branch on that point, only on runtime?

File: application/core/LSYii_Application.php [hunk 2]
```diff
@@ -717,4 +717,93 @@ public function getPublicBaseUrl($absolute = false)
         }
         return $baseUrl;
     }
+
+	/**
+	 * Is REST request
+	 *
+	 * @return boolean
+	 */
+    public function isRestRequest()
```
Comment (olleharstedt) [inline application/core/LSYii_Application.php:726]: Should this rather be a method of the request class? CHttpRequest? Dunno if it's subclassed already.

File: application/core/LSYii_Application.php (hunk 1 above)
Comment (kevin-foster-uk) [reply application/core/LSYii_Application.php]: I noticed we have custom error handling code in the onException() method. I have moved the error handling code in to a separate class and added the extra functionality.

File: application/libraries/Yii/Application/ErrorHandler.php [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+
+use Yii;
+use CWebApplication;
+use CHttpException;
+use LSUserException;
+
+/**
+ * Trait for Application error handling
+ */
+class ErrorHandler
```
Comment (olleharstedt) [inline application/libraries/Yii/Application/ErrorHandler.php:13]: We should probably note somewhere that this class is NOT related to the CErrorHandler component built in in Yii, https://www.yiiframework.com/doc/api/1.1/CErrorHandler

Comment (kevin-foster-uk) [reply application/libraries/Yii/Application/ErrorHandler.php]: I renamed the class to AppErrorHandler. I think this and the namespace makes it clear that this is not Yii code but LimeSurvey code relating to Yii.

**Review by olleharstedt — APPROVED** [review]

**Comment by olleharstedt** [comment]
Approved, but unsure how to test?
